app.py

```diff
@@ -140,80 +140,105 @@ async def create_directory():
     """
-    performs inference on an image, and returns the results.
-    The image and inference results uploaded to a folder in the user's container.
+    Performs inference on an image, and returns the results.
+    The image and inference results are uploaded to a folder in the user's container.
     """
     try:
         data = await request.get_json()
+        model_name = data.get("model_name", "default_model")  # Default to 'default_model' if not provided
         connection_string: str = os.environ["NACHET_AZURE_STORAGE_CONNECTION_STRING"]
         folder_name = data["folder_name"]
         container_name = data["container_name"]
         imageDims = data["imageDims"]
         image_base64 = data["image"]
-        if folder_name and container_name and imageDims and image_base64:
-            header, encoded_data = image_base64.split(",", 1)
-            image_bytes = base64.b64decode(encoded_data)
-            container_client = await azure_storage_api.mount_container(
-                connection_string, container_name, create_container=True
-            )
-            hash_value = await azure_storage_api.generate_hash(image_bytes)
-            blob_name = await azure_storage_api.upload_image(
-                container_client, folder_name, image_bytes, hash_value
-            )
-            blob = await azure_storage_api.get_blob(container_client, blob_name)
-            image_bytes = base64.b64encode(blob).decode("utf8")
-            data = {
-                "input_data": {
-                    "columns": ["image"],
-                    "index": [0],
-                    "data": [image_bytes],
-                }
-            }
-            # encode the data as json to be sent to the model endpoint
-            body = str.encode(json.dumps(data))
-            endpoint_url = os.getenv("NACHET_MODEL_ENDPOINT_REST_URL")
-            endpoint_api_key = os.getenv("NACHET_MODEL_ENDPOINT_ACCESS_KEY")
-            headers = {
-                "Content-Type": "application/json",
-                "Authorization": ("Bearer " + endpoint_api_key),
+        if not (folder_name and container_name and imageDims and image_base64):
+            return jsonify(["missing request arguments"]), 400
+
+        header, encoded_data = image_base64.split(",", 1)
+        image_bytes = base64.b64decode(encoded_data)
+        container_client = await azure_storage_api.mount_container(
+            connection_string, container_name, create_container=True
+        )
+        hash_value = await azure_storage_api.generate_hash(image_bytes)
+        blob_name = await azure_storage_api.upload_image(
+            container_client, folder_name, image_bytes, hash_value
+        )
+        blob = await azure_storage_api.get_blob(container_client, blob_name)
+        image_bytes = base64.b64encode(blob).decode("utf8")
+        
+        # Model function mapping
+        model_mapping = {
+            "Seed Classification": inference.process_inference_results,
```

Comment:
Don’t hesitate to correct me if I’m wrong, but in my understanding inference.process_inference_results objective is to blueprint the box perimeter around the seeds and tag them with the right label the function received from the model data. I thought the model was access with `endpoint_url` and `endpoint_api_key`. Meaning that model_mapping would look like this (in my mind):

```python
model_mapping = {
    ‘Seed_Classification’: (endpoint_url, endpoint_api_key),
    # ‘another_model’: (another_endpoint_url, another_endpoint_api_key)
}
```
Therefore, in the backend, with William we thought that we would need to code a dictionary comprehension with all the endpoint and api key needed. Then we could send to the frontend a `set(model_mapping.keys())` for example and then reworke "/inf" to be able to call all model (or pipeline).

Problem with that is api_key might be exposed and we need to have a way to collect the url and key from a single endpoint (to not have to call all model endpoint just to retrieve them).

Reply:
the concept of API keys are only valid for a server to server call. 

Here the frontend is calling the API endpoint; to secure such an endpoint, you need user authentication and session management.

there isn't a need for different endpoints either as we can use a parameter to /inf to specify the model we want, which gets mapped to a function.

Reply:
Need to distinguish pipelines from models

Pipeline -> one or more model is called
PipelineFunction -> logic to call 

discussed with @MaxenceGui a design using an external model_list.json stored on the blob storage to provide pipeline name, pipeline function, model endpoint(s), model api key(s).

Reply:
## Sequence Diagram first draft

@RussellJimmies, @rngadam feel free to add your comment on what should be modified or update (or have been forget)

```mermaid
sequenceDiagram
    actor Client
    participant Frontend
    participant Backend
    participant Blob storage
    Client->>Frontend: applicationStart()
    par Backend to Blob storage
        Note right of Backend: should be done at deployment
        Backend->>Blob storage: get_pipelines_endpoints_keys(blob_url)
        Blob storage-->>Backend: return pipelines.json
        loop pipelines_mapping
            Backend->Backend: pipelines mapping
        end
    and Frontend to Backend
        Frontend-)Backend: handlerPipelinesNamesRequest()
        Note right of Frontend: a pipeline is the name given to a set of model
        Backend--)Frontend: return set(pipelines)
    end
    Frontend->>Client: applicationReady()
    Client-->>Frontend: ask to change model
    Frontend-)Backend: handlerInferenceRequest(model_name, folder_name, container_name, imageDims, image)
    alt missing argument
        Backend--)Frontend: Error 404 missing arguments
    end
    Backend--)Frontend: return processed_result_json.json
```

Reply:
Here's a overview of where I am with the diagrams and glossary for the inference request. 

I have also prepare doc on markdown file that I can push to this pull request if it's easier to work on.
@CFIALeronB @rngadam @RussellJimmies 

# Inference Request with n pipelines

## Glossary

### Pipelines
Pipelines are define as a set of models that follow each others where output of
one model is used as input for the next models. A pipeline contains from 1 to n
models.

```mermaid
flowchart LR

A{Pipeline entry point}-->|user input|SB1-->F{Pipeline exit point}-->|"last model
output"|G>"return
result.json to user"]

subgraph SB1[In pipeline]
    direction TB

    B(model 01)-->|"model output
    send to the next one"|C(model 02)-->|can have n models|D(n models)

end
```

### Models
A model is a AI model who's part of a pipeline. A model accept images as input
and return json as output. Generally json contains coordinate of object in the
source image, that he may prompt out to feed the next step of the model.


### Model from Frontend
On the frontend interface, a pipeline will be call a model, since the user will
not be aware of the difference. For the user perspective, he sent data to a
model and receive the result.

*Suggestion: we could call the pipeline a method, if we don't want to mix term.*

# Sequence Diagram for inference request

```mermaid
sequenceDiagram

    actor Client
    participant Frontend
    participant Backend
    participant Blob storage
    participant Model

    Note over Backend,Blob storage: initialisation
    Backend-)Blob storage: before_serving()
    Blob storage--)Backend: return pipelines_models.json
    Note over Backend,Blob storage: end of initialisation

    Client->>Frontend: applicationStart()
    Frontend-)Backend: get_pipelines_names()
    Note over Frontend, Backend: HTTP POST req.
    Backend--)Frontend: return pipelines_name.json

    Frontend->>Client: application is ready
    Client-->>Frontend: client ask action from specific pipeline
    Frontend-)Backend: inference_request(pipeline_name, folder_name, container_name, imageDims, image)
    Note over Frontend, Backend: HTTP POST req.
    alt missing argument
        Backend--)Frontend: Error 400 missing arguments
    else no missing argument
        Backend-)Blob storage: mount_container(connection_string(Environnement Variable, container_name))
        Blob storage--)Backend: container_client
        
        Backend-)Blob storage: upload_image(container_client, folder_name, image_bytes, hash_value)
        Blob storage--)Backend: blob_name

        Backend-)Blob storage: get_blob(container_client, blob_name)
        Blob storage--)Backend: blob

        loop for every model in pipeline
            Note over Backend,Blob storage: Need to build endpoint with pipeline_name
            Backend-)Model: urllib.request.Request(endpoint_url, body, header)
            Model--)Backend: result.json
        end
        
        par Backend to Frontend
            Backend--)Frontend: return processed_result_json.json
        and Backend to Blob storage
            Backend-)Blob storage: upload_inference_result(container_client, folder_name, result_json_string, hash_value)
        end
    end
    Frontend--)Client: display result
```

![footer_for_diagram](https://github.com/ai-cfia/nachet-backend/assets/96267006/cf378d6f-5b20-4e1d-8665-2ba65ed54f8e)

Reply:
Backend-)Blob storage: before_serving()

Reply:
@MaxenceGui suggest moving this conversation to the issue #51 as we want to capture design in the issues.

Reply:
Done it has been move to issue #51